### ISSUE TITLE
Proposal: Support for Cross Browser Testing

### ISSUE THREAD

**Issue by brian-mann** [cypress-io/cypress#310]
# UPDATES: February 7th, 2020

📦 We have added support for Firefox and MS Edge in **Cypress v4.0.0**.
- [📣 Read Announcement](https://www.cypress.io/blog/2020/02/06/introducing-firefox-and-edge-support-in-cypress-4-0/)
- [📘 Read Cross Browser Testing Guide](https://on.cypress.io/cross-browser-testing)
- [📋 Changelog](https://on.cypress.io/changelog#4-0-0)
- [🛠 Migration Guide](https://on.cypress.io/migration-guide#Migrating-to-Cypress-4-0)

# UPDATES: Oct 20th, 2018

I gave more updates on the work we've been doing for cross browser support and other features on a webinar we gave.

https://youtu.be/FfqD1ExUGlw?t=2079

# UPDATES: Sept 8th, 2018

Cross browser testing is now a front and center major focus here at Cypress. We've been experimenting over the last few months and have a working implementation to add native event support for Chrome, Firefox, Edge, and IE11. We also believe adding Mobile Safari and Android should also be straightforward - with Safari actually being the most difficult to get.

The architecture by which we are doing cross browser testing is a best-of-breed design where we tap into the native API's for Chrome and Firefox (avoiding the use of webdriver) and for Edge and IE11 we only use a minimal set of Webdriver API's - none of which are the traditional `sendKeys` or `value` endpoints, which avoids the network latency problem entirely. We plan to eventually publish this architecture so people understand how Cypress is different than Webdriver. None of the other core components of the Cypress architecture are affected by this change, it is simply how to enable native events in a consistent manner across all browsers without introducing the latency / flaky / non-determinism of Webdriver.

In additional to all these things, we are also adding Sauce Labs support (at a minimum) and will also have a proof of concept prepared for that.

As we add support for native events we are subsequently adding support for the browsers at the same time to ensure all of our algorithms are normalized correctly. However, you'll likely first see us release native event support in Chrome + Firefox, followed by IE11 sometime after that.	

...below is the original post from 2016...

## Does Cypress support cross browser testing?

Currently the answer is: **No**

Right now, we only support Chrome* variants such as `Chrome`, `Chromium`, and `Canary`.

## Will you ever support other browsers?

**Yes.**

We will be able to support other Desktop Browsers and Mobile Browsers (using Webviews).

We have built Cypress from day 1 with cross browser testing in mind, and are very conscientious to ensure the API's we write today will work in the future.

# The Technical Details

## Why do you only support Chrome today?

- Cypress taps into the underlying browser Automation API's of Chrome to do things like modify cookies, take screenshots, focusing the browser window, etc.
- We also launch the browser using very specific command line flags to turn off things that get in the way of automated testing.
- We have synchronized these Automation API's when Cypress runs headlessly (in Electron). This means running headlessly behaves identically to a real headed Chrome browser.

## What do you mean by "Automation API's"?

- Automation API's describe the functions that a browser exposes to do things outside of the **Javascript Sandbox**.
- Currently Cypress comes with its own universal **Driver** which implements most of the `cy` commands using Javascript - in the same run loop as your application. This is superior in the sense that Cypress can understand, know, wait for and modify its behavior based on how your application reacts to its commands in real time.
- The downside is its impossible to fully automate a browser entirely in Javascript. Javascript is a security sandbox which restricts what you can do.
- To fully automate a browser you must extend beyond the sandbox and tap into the underlying Automation API's which do not have these restrictions.

## What will it take to support cross browsers?

- We have to port commands which need to extend outside of the sandbox to each other browser we are attempting to support.

## How will you port these Automation API's?

- Either manually by hand
- Or by using the existing browser drivers

## What do you mean by "existing browser drivers"?

- Each browser has built its own driver to the [Webdriver Spec](https://www.w3.org/TR/webdriver/).
- Here are some examples of drivers:
- Google Chrome Driver (ChromeDriver)
- Micosoft Edge Driver
- SafariDriver
- Mozilla GeckoDriver (Firefox)
- Internet Explorer Driver
- Appium

## Why would you port these by hand?

- Because using the existing drivers comes with a host of problems.

## What's wrong with the existing browser drivers?

- They are built for and thus limited to the Webdriver spec.

## What is the Webdriver spec?

- It's a HTTP API spec which defines a common interface to drive all browsers consistently.
- It enables developers to abstract away the details of how each different browser implements its Automation API's

## What is the problem with Webdriver?

- It's slow
- It's async
- It's stateless
- It's exposed over an HTTP server
- It's a lowest common denominator
- It's limited in scope and features
- There are small inconsistencies
- There are various driver bugs

## What do you mean by "lowest common denominator"?

- The spec is essentially limited to only including "all the things that every browser can do".
- This leaves a tremendous amount of *power* and capability off the table that more modern browsers provide.

## Can you give me some examples?

- In Webdriver you cannot clear all cookies for all domains - you can only clear cookies for the current domain context under test.
- There is no way to tell the browser to clear its cache.
- Chromedriver [cannot run tests with Dev Tools open](https://sites.google.com/a/chromium.org/chromedriver/help/devtools-window-keeps-closing).
- Oftentimes the drivers may [miss elements](https://sites.google.com/a/chromium.org/chromedriver/help/clicking-issues) you're trying to interact with.
- You cannot easily control the application under test.
- You don't have native access to `window`, you're limited to stringifying objects over the wire.

## Does Cypress currently use Webdriver?

- No. We wrote our own universal driver.

## What would be the benefit of using Webdriver?

- We could possibly use the existing drivers to port over various commands without having to write the manual integrations by hand.
- This would allow us to support cross browsers much faster.

## If you use Webdriver won't Cypress also inherit these limitations?

- In *most* cases - **no**.
- In *some* cases - **possibly**.

## How can you avoid these limitations?

- By using our universal driver to do as much automation as possible and only go outside of the browser for a limited set of commands.
- The whole point of Cypress is to exceed the boundaries of the current testing ecosystem.
- Because we automate the browser with a completely different architecture we are in a position to extend beyond the Webdriver Spec and also normalize its behavior consistently across all browsers.

## Can you give me an example of this?

- Cypress can [clear all cookies of all domains](../issues/408) instead of just the current domain under test.
- We perform work outside of the browser but then synchronize it seamlessly like with [cy.request](https://on.cypress.io/api/request).
- We have many commands which work outside of the browser which don't need to be ported.

## But won't you still be limited by Webdrivers flakiness?

- No. Even if we were to issue native events through Webdriver, because your application is always tested *through* Cypress, that means Cypress can respond synchronously to events that take place, and modify its behavior accordingly.
- The vast majority of commands do not ever have to leave the browser anyway.

## Can you give me an example of what you're talking about?

- Imagine this scenario: you have a modal opening which animates in, and you want to click the "Submit" button.
- In Webdriver, because its an HTTP API which often involves multiple network hops, there is usually a significant delay.
- Let's assume you're attempting to issue a native "click" to the submit button.
- When you issue the "click" command, the browser driver calculates the center coordinates of the button at say: `{clientX: 50, clientY: 100}`.
- It then issues the native click event with those coordinates.
- By the time your application "receives" this event, it could possibly "miss" the button altogether.
- Your test fails because your button was never clicked, and the modal was never dismissed.
- What happened?
- Because the modal was animating it, its coordinates were changing rapidly. Webdriver has no idea or concept of this, it only takes a coordinate snapshot at the moment in time and because it asynchronously dispatches the native event, it misses.

## Why isn't this a problem with Cypress?

- Because Cypress is in the same run loop as your application, its impossible for it to "miss".
- It can synchrously run calculations, and it already automatically figures out if an element is animating beyond a certain threshold - and therefore wait until it stops.
- If it did issue a native event, it could detect that this event has yet to happen, and if the element moved from the coordinate, immediately attempt to recover by either canceling the click event or issuing another click at new coordinates.

## When do you think you'll support other browsers?

- We really don't know, its not a priority for us right now.

## Why isn't it a priority? This is super important!

- We disagree, cross browser testing is **not as important** as you think it is.

...

### The reasoning and examples behind this are a WIP. We'll be updating this doc when they are ready.

**Comment by ls-guillaume-lambert** [comment]
Hey @brian-mann!

First, amazing job on Cypress. I've been researching the best tool out there for JavaScript e2e testing and Cypress is up there with TestCafé for me. The only thing really that prevents me to go full on with Cypress is the cross-browser testing capabilities. 

Looking at the above comment, I understand that Cypress does not consider cross-browser testing to be important. Is there anything anywhere that can convince me that I should be thinking the same way? Various OS/Browser combinations can yield different results, and I still think the value of automated cross-browser e2e testing is valuable. 

In any case, this is really the only thing that prevents me from choosing Cypress as my weapon of choice for e2e testing, and I wanted to have you updated opinion on this since this issue was opened about 6 months ago.

Is there a way we can talk about this through DM? I'd really love to discuss this with you or any Cypress team member.

Thanks in advance and keep up the amazing work!

**Comment by ls-guillaume-lambert** [comment]
For those interested in knowing how the Cypress team/devs feel about this, head over to there [Gitter chat](https://gitter.im/cypress-io/cypress) and search for "cross-browser". @brian-mann took the time to give me a lot of details on this, and he also told me he would update this issue to provide more context / examples / more compelling argument.

**Comment by shcallaway** [comment]
@ls-guillaume-lambert Thanks for sharing! I had some similar questions.

**Comment by shcallaway** [comment]
@brian-mann 

> anyway we will be adding other browsers
> firefox and edge

Could you provide any updates on support for these browsers? Is it in development? Also, a few people mentioned support in CI for other browsers via Sauce Labs and/or Browserstack. What's the status there? 

Love Cypress! :)

**Comment by tonypee** [comment]
Seconded, an update on this would be great! thanks

**Comment by brian-mann** [comment]
It's the next project one of our team members is going to be working on. We have the implementation plan ready to go.

**Comment by brian-mann** [comment]
Nope sorry its gonna be @chrisbreiding 

You gotta finish parallelization and load balancing ;-)

**Comment by joegaffey** [comment]
Cypress looks fantastic but without cross browser support I'll have to stick with Selenium. 

Firefox then Edge would be my preferred priority. 

Any possibility for iOS Safari support?

**Comment by vvo** [comment]
> We disagree, cross browser testing is not as important as you think it is.

I am also interested in having more details about this. I don't think unit testing for cross browsers is important. But for e2e I always thought that yes this was important.

**Comment by brian-mann** [comment]
We will be adding cross browser support likely by EOY. @chrisbreiding next project is that. 

The support will be:

- Firefox
- Safari
- Edge

No IE support is planned. It is substantially more difficult.

**Comment by brian-mann** [comment]
We are planning on writing up a full blog post on what we think e2e testing should look like, and how cross browser fits into that. It's going to require some pretty graphics ;-)

**Comment by danieldram** [comment]
This is an amazing solution, but without IE, FF and Safari support, I can't get buy in from my office to use this :(

**Comment by joelgrimberg** [comment]
is connectivity with SauceLabs/CrossBrowserTesting/Browserstack also considered?

**Comment by nrxsmith** [comment]
Do you guys have an update on your progress with this? Any ETA?

Thanks!

**Comment by brian-mann** [comment]
@chrisbreiding next project is this - starting this month. He's finishing up the thing he's been working on now.

**Comment by jennifer-shehane** [comment]
Support for Firefox browser proposal: https://github.com/cypress-io/cypress/issues/1096

**Comment by makenova** [comment]
@brian-mann you [mentioned a blog post](https://github.com/cypress-io/cypress/issues/310#issuecomment-337349842) in this thread.

> We are planning on writing up a full blog post on what we think e2e testing should look like, and how cross browser fits into that. It's going to require some pretty graphics ;-)

May I have a link to it?

**Comment by green-arrow** [comment]
@brian-mann - I’m also very interested in a blog post about the rationale behind cross-browser testing not being as important. I read a few comments in the gitter channel about how modern browsers largely resolve a lot of cross-browser bugs, but I just don’t buy it. There are still huge inconsistencies between browsers with modern features like flex box. Chrome, Firefox, and Safari each have their own quirks with just this one feature that our current cross-browser test suite catch. Even more seemingly benign issues like z-index creating stacking contexts differently in every browser can cause the application to break for a user. 

A blog post that addresses some of this would definitely be welcome.

**Comment by brian-mann** [comment]
Yeah unfortunately, we never wrote it. I haven't necessarily changed my position, which is that repeating the same identical test suite across a variety of browsers is not the way to go - but regardless we are adding cross browser support right now. For users wanting utter and complete coverage, you'll have that ability.

There is an law of diminishing returns here where each successive browser only marginally increases your confidence - and requires substantially more infrastructure and cost. It may be valuable in some cases to some users, but overall we run millions of tests, and only a small fraction of those failures occur in other browsers which would not have been caught by automated functional tests. I believe that's the key here - functional tests will only ever catch a specific class of bug that is mostly **logic** related, and oftentimes not necessarily visual.

The bigger solution and really the only one that really solves cross browser issues for good is visual regression testing (screenshot diffing). We are also adding support for that as well. With that feature, I would then recommend full testing in chrome, but only marginal happy path testing in the rest of the browser with screenshot diffing. This balances what is oftentimes the primary root cause of cross browser testing (visual issues), while still covering the functional aspects, but primarily only in one browser.

**Comment by NoNameProvided** [comment]
Hi @brian-mann!

No rush or anything, buy do you have maybe a date to share? We are starting a new project soon and would love to use Cypress for testing, but support for Firefox/Safari/Edge is a requirement for us.

**Comment by brian-mann** [comment]
Trying to get Firefox done this month. Maybe another browser per month after that.

**Comment by serhiibr** [comment]
Hi @brian-mann!
Do you plan to support Opera browser in the foreseeable future?
Thank you in advance.

**Comment by brian-mann** [comment]
First passing test running Cypress in Firefox...

 
![screen shot 2018-01-09 at 11 13 11 pm](https://user-images.githubusercontent.com/1268976/34789050-bb825a32-f60a-11e7-8c09-da2f434d1729.png)

**Comment by brian-mann** [comment]
@serhiibr if there is demand for it we'll add it... but currently its behind Firefox, Edge, and Safari.

**Comment by Mouvedia** [comment]
> Do you plan to support Opera browser in the foreseeable future?

By that I hope you meant Opera 12.  
[OperaPrestoDriver](https://github.com/operasoftware/operaprestodriver) would be a good place to start from.

Opera 15+ is blink-based (like Chrome) so it's not a priority at all IMHO.

**Comment by Mouvedia** [comment]
> but currently its behind Firefox, Edge, and Safari.

@brian-mann Safari is the new IE 6; I hope you will pick it up before Edge—which is pretty much compliant.

**Comment by serhiibr** [comment]
@brian-mann  Thank you. I agree that Firefox, Edge, and Safari are more vital.
@Mouvedia Yes, I meant Opera 12, thank you.

**Comment by FelipeTaiarol** [comment]
Great work with Cypress! Also trying to make a case to use it instead of Selenium but the lack of IE and Safari are making it harder to sell. I hope you guys are able to add cross browser support soon :)

**Comment by EirikBirkeland** [comment]
I'm so excited about Cypress.io! Looking forward to more browsers soon to avoid having to use Selenium :D Firefox, Safari and Edge should work for me. IE11 would be nice as well, but it's better to encourage users to stop using unsupported browsers with potentially unadressed security holes.

**Comment by iGitScor** [comment]
Hi, Appium support would be awesome and CypressIO would be a master reference for E2E testing also in hybrid apps world ❤️

**Comment by maccurt** [comment]
I for one would like to see IE die a fast death. The people who pay me money also balk at the lack of IE support, but with what I have seen so far with this tool I feel confident the ease of use outweighs the concern for IE

**Comment by pgroot91** [comment]
Not sure but will send here some feedback while testing with **Canary** and **Chromium** witch are both version 66. I know those are buggy and loaded with more expermental features. My screen is tripping shaking like crazy when there are type actions in input fields is this a normal behavior or is this something caused in Cypress itself running it in GUI mode? Anyway looking forward to see more supported drivers/browsers coming to Cypress.

**Comment by rssfrncs** [comment]
Is there a rough estimate/goal for Edge/IE support? We are in the process of selecting a test framework at work and this is the only reason to not use Cypress. We'd be happy to get on board with Cypress if we knew that in **x** months Edge/IE would be supported.

**Comment by EirikBirkeland** [comment]
Since corporations may still use IE11 for another few years, it may make sense to support it. I believe this happens when the computers run Win7 with IE11 or something like that. However ridiculous.

**Comment by EirikBirkeland** [comment]
Another route is to skip IE11 and only support it if it still makes sense after having added support for all current browsers.

**Comment by rssfrncs** [comment]
@EirikBirkeland where i work windows 10 is the default os, however ie11 is still the default browser for intranet sites and not edge!

**Comment by EirikBirkeland** [comment]
@rssfrncs I thought IE11 had been discontinued, but I was mistaken.

> Windows 10 will be mainstream supported until October 13, 2020, and the end of extended support is October 14, 2025, **you’ve likely got until 2025 before they’ll stop supporting IE 11, even though they’d prefer you to use Edge**.

**Comment by laurentpayot** [comment]
I was surprised to see that IE 11 is still more than twice more used than Edge: https://netmarketshare.com/browser-market-share.aspx?id=browsersDesktopVersions

**Comment by brian-mann** [comment]
We are first supporting Firefox. Edge is also the next easiest - but we've internally talked about adding Mobile Safari + Android first before Edge.

IE + Safari are definitely the hardest. Safari is difficult because their safaridriver is not open source and has security limitations that would prevent actually being able to "use" Cypress while automated tests run. IE is also difficult because well Windows automation is horrendous and IE requires manual steps for security cert acceptance, which is also a problem.

Both of these likely "can" be supported at some point, but it would likely require significant development resources (which is fine, we're a funded startup + hiring) but I can't even begin to approximate a date.

**Comment by iGitScor** [comment]
Glad to hear that Android is one of the priority 💯

**Comment by cavaloni** [comment]
I love this tool. You guys have done great work. I'm having a hard time convincing my team, unfortunately, due to this lack of support. Looking forward to the reasoning and examples to hopefully sway them in the right direction.

**Comment by adamchenwei** [comment]
Here as well. A bit tough to have a full buy in because of this. Would be nice to have a timeline on all browser support or reason not to or never going to will be great.

**Comment by zbarbuto** [comment]
The reason it's not a priority has been listed has already been posted in the thread https://github.com/cypress-io/cypress/issues/310#issuecomment-354925454

Firefox support is already under way https://github.com/cypress-io/cypress/issues/1096 with an open pull request: https://github.com/cypress-io/cypress/issues/1096 - I suggest subscribing to that issue for status updates.

**Comment by kunwardeep** [comment]
I really have massive amount of appreciation for cypress. But unfortunately companies do need to support IE 10/11. The reason for this is that their clients use such browsers because of which this needs to happen. I don't think there are massive differences in functional testing between browsers, its mainly visual. We run visual tests across browsers but not functional. The problem i face is that should i maintain two suites, one using cypress for functional vs one using selenium to do multi browser. Hence unfortunately it becomes that we end up using selenium

**Comment by rhymes** [comment]
@kunwardeep [TestCafe](https://testcafe.devexpress.com/) supports multiple browsers

**Comment by aecz** [comment]
@EirikBirkeland Good points except for the graph.
It shows that Chrome should indeed by the first supported browser but we can see that in the past 2 years, there is no significant decrease in Internet Explorer share and that it is tied to second place with Firefox for Desktop/Laptop browsers.
I understand Firefox comes before IE as it is easier to implement as it was said above.
I could understand Safari before IE because its has a bigger share on Mobile although we usually need to test more on IE than Safari once Chrome is tested.
I cannot understand Edge before IE :(

**Comment by EirikBirkeland** [comment]
@aecz I see your poitn. And I would add that it's a fact that some of the IE11 'users' are large companies with a ton of users. So, what if our end-clients are all using IE11? In this case, the graph above means even less.

**Comment by corysimmons** [comment]
@kunwardeep IE is only 3% users and dropping. Those users are probably extremely used to having a sub-optimal web experience. I hope the Cypress team doesn't waste time supporting IE (since they already mentioned it was "substantially" more difficult.

Just do a quick manual visual test every now and then? Or, if you have a huge system that needs automating, and your audience is primarily stuck on IE, perhaps use something like BrowserStack/Saucelabs and integrate screenshot diffing. If anyone goes the screenshot diffing route, please open source. :)

**Comment by jescalan** [comment]
@kunwardeep is absolutely correct. And unfortunately, the percentage is different for each site depending on their audience. I have worked with clients who have sub-1% IE traffic and clients who were well over 15%. The global average percentage of users on IE just isn't enough to make a definitive statement on whether it's "worth it" to do something.

Even imagining you couldn't track specific stats on who is converting on your site, and which browser they are on, which surely just about everyone is able to do, you can get a rough measure by running some very basic multiplication -- percentage of IE users by monthly total users by conversion rate by average sale - there you can get a direct dollar value on how much you'd lose if an IE error were to stop users from converting. For some companies, this might be so low it's not worth spending time to address. For others, it could be very much the opposite. For large sites (like amazon, for example) that number is surely well into the millions of dollars, and is unquestionably worth keeping a close watch on.

Arguing that testing in IE is not worth it for any company is absurd. For some companies it is, and for others it isn't - there's no universal argument here. Do your own research and make a decision for your company that is based on easily accessible analytics. What this thread really boils down to is whether it's a good business decision for cypress, a for-profit company, to spend their time building out IE support, based on how much time it would take them and how much their customers would pay for this feature. By the company representatives' responses in this thread, it seems like the answer right now is "no". It might be something that re-appears on their radar eventually as they clear out higher priority tasks, and when it does, they will surely notify the many users who are interested in the feature here ✨

**Comment by corysimmons** [comment]
Hi dad :)

**I** hope some old company doesn't offer Cypress a million dollars and force **me** to wait for sweet features. 😾

To be clear, I don't hate IE support. I was just stating a personal opinion as someone who fought against the tide for years, and missed out on a lot of cool stuff while doing that. I don't want the Cypress team to be lulled into the same trap unless Amazon is offering them millions of $ right now.

The more interesting (vs the ancient debate on when we're allowed to stop supporting Microsoft browsers—which I'm done talking about), and overlooked, point I made was pushing for screenshot diffing.

Cypress (or anyone) could support every browser via screenshot diffing (if it's performant enough). Then they could focus strictly on making Cypress cleaner/easier to use, and consolidate these 20 "Support Firefox", "Support XYZ" features down to 1 feature.

Screenshot diffing could also double as a design tester (i.e. if it looks good in Chrome at the various resolutions your original test suite is running in, it will look good at those resolutions in everything else at those various resolutions).

There will be very slight rendering differences between browsers/devices, fonts and such, but a threshold setting could address that. Anything that broke out of that threshold, you could assume is mis-rendering in whatever device and fix.

At the very least, this could serve a _lot_ of people's needs while serving as a bandage until they are able to add official support for every browser/device.

**Comment by matthewblott** [comment]
Just my tuppence worth as a former Windows programmer as I see a bit of heat regarding IE support. I come down on the side of the Cypress' team decision to not offer support for IE. Windows XP is no longer supported and Microsoft offered free upgrades to Windows 10 from Windows 7. Windows 10 ships with Edge and this is Microsoft's main focus. In other words there's no reason Windows users (particularly business customers) can't be using Edge. Unlike IE, Edge is a true evergreen browser and more compatible with ES6 and other modern features. IE has improved a lot but it is now legacy technology. I spent considerable effort trying to get my .NET Selenium tests to run with IE and I had to abandon the effort (with workarounds I got all the others working). I can't see any compelling reason the Cypress' team would need to go down that painful route.

**Comment by SaltyDH** [comment]
The flip side to this is IE support would allow automation to cover the cases that no one else can. No one *willingly* uses IE to develop an application or to test it (right?). Having IE support in Cypress would provide coverage over one of the browsers most likely to regress due to an ill conceived change that works on all the evergreen browsers just fine. For us, its one of the most important gaps in our current testing strategy - we already know some clients will use IE11 exclusively. I'm still trying to wrap my head around all the necessary pieces that make Cypress compatible with a specific browser.

@brian-mann Do you have any suggestions on how I might come up to speed on this if I was to try and tackle an IE Cypress driver?

**Comment by danieldram** [comment]
At first I was wondering why not support IE and it really makes sense to me that they don't support IE. The more we keep creating code to work on IE the more people will find it useful and refuse to upgrade their browsers. We as developers know that there are better solutions that are also free for the consumer, they should upgrade.. Perhaps making it more difficult for developers to develop for those browsers by limiting scope of tools coverage is a great way to finally let IE finally be a thing of the past. Businesses want to support IE but if there is no way produce in those environments because our dev tools don't cover it, it forces them to think how to convert users to modern standards. just my opinion

**Comment by SaltyDH** [comment]
I agree with the perspective above, my reality is we have long term contracts in place with our clients who in turn have contracts with their clients to support IE11 while it is still supported by Microsoft. Additionally our competitors still support IE11. I'm all for continuing to educate and being part of the solution but I also would like to put food on the table :)

**Comment by lucadonnaloia** [comment]
+1 following

**Comment by adamchenwei** [comment]
all you need to do is click subscribe button on the right top of the thread @lucadonnaloia

**Comment by BrowserTechnicalSupportNumber** [comment]
Get in touch with [Browser Technical Support](http://www.browsertechnicalsupportnumbers.com) at 1-800-240-2551 and resolve your all issues related to google chrome, Mozilla Firefox, Apple Safari and Internet explorer.

**Comment by Alon87** [comment]
@brian-mann & @Cypress team. Can we get an expected date when Cypress.io will start to support Firefox, Safari, Edge etc? Will there be a beta version for us to try?

**Comment by Herteby** [comment]
Personally, IE support would be more important than Firefox, Edge and Safari combined, since IE is the one most likely to diverge.
24% of our users are sadly still on IE.
Even among Windows 10 users, IE is more common than Edge.

**Comment by dn1231** [comment]
I really like Cypress but for now I have to use testcafe and the main reason is the browser compatibility.

**Comment by emijmker** [comment]
I completely agree with @Herteby so here's another upvote for IE to be added over other browsers. Our experience is that IE is often off and causing more bugs than other browsers. Chrome/Firefox/Safari can work properly, but it's always IE that has to be different. Would save a lot of time to have Chrome and IE tested automatically, that would provide sufficient coverage and take the very rare exception for granted where FF shows a bug that doesn't come up in Chrome. I'm not a fan of IE, in fact I hate it. But there are still so many IE users whether we as techies want to or not and you can shout all you want, it's hard to ignore. The fact that IE is harder to add than FF IMHO should not be a reason for it to have lower priority, it should be about where are the mosts risks.
Love Cypress so great work on that guys! So in regards of IE we're stuck to manual testing it, but love to see this change.

**Comment by jcollum** [comment]
So in the meantime what are people using for testing of Chrome & Safari & Edge & Firefox in one tool?

**Comment by pekalang** [comment]
We are using Testcafe.
Even if you have to get used to the syntax.

**Comment by Herteby** [comment]
@jcollum We're using Robot Framework/Selenium

**Comment by subbiah28062** [comment]
Hello Dev,

I would like to know whether you are extending support to safari.

I came across browser-launcher2. can we make cypress to run test in safari using browser-launcher2??????

Thanks
Subbiah

**Comment by brian-mann** [comment]
I added updates for cross browser support here: https://github.com/cypress-io/cypress/issues/310#issue-191595673

TLDR; we are adding cross browser support as a major priority here. Work has been going on for months and we're ready to begin implementation.

**Comment by matthewblott** [comment]
Thanks for the update. My only question would be is IE11 support being prioritised over Edge? Microsoft are advising people to move to Edge so I would have thought this would be a higher priority.

**Comment by c198130** [comment]
Super exciting news.  We're looking forward to further announcements.

**Comment by Mouvedia** [comment]
Without taking the implementation difficulty into account, here's a list ordered by bug proneness in relation to their standard compliance:

1. Opera 12
2. Safari
3. Internet Explorer
4. Edge and Firefox (ever green)

Concerning usage share—depending on your source—either Firefox or Safari is the runner-up behind Chrome. Opera 12 is dead last in that regard.

**TL;DR**

You got to support first and foremost Safari and Firefox. Internet Explorer comes second. Edge and Opera 12 when you are done with the first 3.

**Comment by davidjeddy** [comment]
I mean...

![image](https://user-images.githubusercontent.com/6232455/45691351-6b691400-bb26-11e8-8d2c-6274463168d1.png)

Source: https://tinyurl.com/y7bahcg2

**Comment by Mouvedia** [comment]
> **depending on your source**

http://gs.statcounter.com/ and https://en.wikipedia.org/wiki/Usage_share_of_web_browsers got Safari second for example.

**Comment by hdjong1** [comment]
You guys are comparing desktop and mobile market shares. When looking at desktop only, Safari's share is very low. When factoring in mobile, Safari suddenly makes a huge jump because all iPhones and iPads also run Safari.

**Comment by matthewblott** [comment]
Agree with @hdjong1, Safari usage is very low in the desktop world. Devs probably think it's higher coz a lot of them are Mac users.

**Comment by Mouvedia** [comment]
More than 50% of the trafic is done on mobile. The browser engine on Apple devices is locked so even if you are using Chrome internally it's just JavaScriptCore or Nitro. The problem is that unless you update your version of iOS you cannot upgrade your browser. Hence why Safari is considered "the new IE".

**Comment by mike-nelson** [comment]
Happy to see IE11 support being a priority. It’s still huge in the corporate world and harbours more differences to chrome that make it worthwhile testing on. 

I am also thinkiing it would be cool if you could effectively test older versions of IE by using IE11’s ability to emulate old browsers. This would be amazing for testing old browser compatibility. Although supporting old browsers is not generally considered cool by developers it’s actually pretty snobbish towards the other half of the digital divide - older and poorer people who still have an old computer.  We can’t justify manually testing on every version of IE but automating it could help make this coverage more realistic.

**Comment by jahmed2345** [comment]
Really happy to see that people are now acknowledging the importance of cross browser testing for making their website pixel perfect. Users willing to test their website have an easy way to do it now. Just visit [https://www.lambdatest.com/] and start your website testing. Happy to help!

**Comment by SaltyDH** [comment]
Bad form, a service I will avoid.

**Comment by toshovski** [comment]
Currently we are evaluating cypress and it is really a great framework. We had a discussion today about cross browser functionality. Do you guys plan to add new browsers by the end of the year? Is there a more detailed road map? We have mostly corporate customers. We need to make sure, that Firefox, IE and Edge are also working.

**Comment by jennifer-shehane** [comment]
Brian Mann (@brian-mann) gave some updates on the work we've been doing for cross browser support as well as other future features on our recent webinar - so check it out if you have your eye on this feature: https://youtu.be/FfqD1ExUGlw?t=2079

**Comment by toshovski** [comment]
Thank you @jennifer-shehane for sharing the link. It kind of answered my questions. Unfortunately there is no timeline about releasing the multi browser support, but it looks like you have a lot to do before the release.

**Comment by serkanyersen** [comment]
@jennifer-shehane Are you going to wait for all browsers to be ready or will you release as soon as one browser is ready and keep adding browsers on the way?

**Comment by jennifer-shehane** [comment]
@serkanyersen There is some minimal viable work that has to be done to account for all browsers first, but no, when we feel another browser is ready, we will release that.

**Comment by adamchenwei** [comment]
> > We disagree, cross browser testing is not as important as you think it is.
> 
> I am also interested in having more details about this. I don't think unit testing for cross browsers is important. But for e2e I always thought that yes this was important.

I am super curious about this justification as well, good job so far though, Cypress! Looking forward for the final hockey stick on adoption once this ticket is closed! And ... before that, I am just like anyone else, eye balling on what ppl think about this topic out there...
https://www.quora.com/unanswered/Why-cross-browser-testing-is-not-important

**Comment by willischu** [comment]
> > We disagree, cross browser testing is not as important as you think it is.
> 
> I am also interested in having more details about this. I don't think unit testing for cross browsers is important. But for e2e I always thought that yes this was important.

Agreed that unit testing for cross browsers is unnecessary! Having a small **smoke test** suite specifically for testing across different browsers would be ideal.

**Comment by irfancharania** [comment]
@brian-mann: wrt the comment around "[WebDriver API lowest common denominator](https://github.com/cypress-io/cypress/issues/310#issue-191595673)".
> The spec is essentially limited to only including "all the things that every browser can do".

How does Cypress get around this as you add support for other browsers?
If a browser doesn't have a specific feature, how do you address that?

Would you end up in a scenario [like iMacros](https://wiki.imacros.net/Command_Reference) where each command lists browsers it will work with?

**Comment by irfancharania** [comment]
> IE11 differs hugely from other browsers and occasionally breaks in weird ways that we can't forsee. It would be definitely worth supporting it.

They demo'd a [working proof of concept here](https://youtu.be/FfqD1ExUGlw?t=2620)

**Comment by kunwardeep** [comment]
Another ray of hope https://www.windowscentral.com/microsoft-building-chromium-powered-web-browser-windows-10

**Comment by Mouvedia** [comment]
Edge is evergreen, IE isn't. This news doesn't change a thing.

**Comment by UncleGus** [comment]
YAY! Microsoft are abandoning Edge!

But BOO, they're building YET ANOTHER BROWSER.

Is there any way we can make them just stop?

**Comment by divyaarjnani-narvar** [comment]
Does cypress support cross browser testing am confused by reading so many comments

**Comment by divyaarjnani** [comment]
Does Cypress supports firefox and IE, got confused by reading the comments.

**Comment by highvoltag3** [comment]
@divyaarjnani it does not currently, it is planned for the near-ish future (there's no set date or ETA). This issue has all the updates at the top the comments are made by anyone and has opinions, questions, suggestions and whatever so refer to cypress docs for supported functionality or follow this if you'd like to keep an eye on how things develop/evolve.

**Comment by nguyenj** [comment]
Any plans on integrating it with saucelabs/browserstack/or a similar service? I know this has been asked before; but I can't seem to have any responses that directly answers that question.

**Comment by kellyprankin** [comment]
Can anyone elaborate on this statement:

Why isn't it a priority? This is super important!

    We disagree, cross browser testing is not as important as you think it is.

If I want to convince people to adopt this tool, I feel like they are going to scoff at that statement. Can you give any ammo?

**Comment by jennifer-shehane** [comment]
@kellyprankin @sbasinger The comment on the importance of cross browser testing is an older comment. Our new stance:

>Cross browser testing is now a front and center major focus here at Cypress. 

We're working on support. It is a major undertaking and dependent on implementing [Native Events](https://github.com/cypress-io/cypress/issues/311) first, which is making progress every day.

**Comment by adamchenwei** [comment]
Still having a hrad time to sell Cypress to my team and company. The giantic piece of ugly hole is missing support of other browsers, I mean, if its IE, we can work with it, but not even firefox and safari is a bit tough... really looking forward to a fully browser supported version ... and when that did come, would be nice to come with a selenium test to cypress test convertor ha.

**Comment by Gpx** [comment]
Hi everyone, sorry for spamming. I follow this issue because I want to know when there are updates on cross-browser support. I believe many others do the same.

Instead, I receive notifications about people saying that they need Cypress to work on this or that browser or asking when will support land.

Can I please ask the maintainers to lock comments for outside members so that only relevant information gets shared? If that's not possible it's ok; I'll find another way to keep up-to-date with Cypress.

Sorry again for the noise and thanks to the maintainers for their hard work on Cypress!

**Comment by jennifer-shehane** [comment]
We are actively working on releasing Firefox support. Currently the release is blocked by a garbage collection issue that causes Firefox to crash during some tests during `cypress run`. We're working with the Firefox team and also trying several creative workarounds to prevent the crashing. https://github.com/cypress-io/cypress/pull/1359

We are also working on releasing support for Microsoft Edge as part of this PR. https://github.com/cypress-io/cypress/pull/6181

They are currently targeted for the same release.

**Comment by amirrustam** [comment]
📦 We have released support for Firefox and MS Edge in **Cypress v4.0.0**:
- [📣 Read Announcement](https://www.cypress.io/blog/2020/02/06/introducing-firefox-and-edge-support-in-cypress-4-0/)
- [📘 Read Cross Browser Testing Guide](https://on.cypress.io/cross-browser-testing)
- [📋 Changelog](https://on.cypress.io/changelog#4-0-0)
- [🛠 Migration Guide](https://on.cypress.io/migration-guide#Migrating-to-Cypress-4-0)

**Comment by amirrustam** [comment]
Now that Cypress supports [Firefox](https://cypress.io/blog/2020/02/06/introducing-firefox-and-edge-support-in-cypress-4-0/) in addition to Chrome-family browsers (including [Edge](https://cypress.io/blog/2020/02/06/introducing-firefox-and-edge-support-in-cypress-4-0/) and Electron), this issue is being closed in favor of dedicated proposal issues for each browser under consideration.

- macOS Safari Support: #6422 
- IE 11 Support: #6423 

Please direct further commentary and feedback within these dedicated issues.
